pkg/smokescreen/acl/v1/acl.go

```diff
@@ -0,0 +1,230 @@
+package acl
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/sirupsen/logrus"
+)
+
+type Decider interface {
+	Decide(service, host string) (Decision, error)
+}
+
+type ACL struct {
+	Rules            map[string]Rule
+	DefaultRule      *Rule
+	GlobalDenyList   []string
+	GlobalAllowList  []string
+	DisabledPolicies []EnforcementPolicy
+	*logrus.Logger
+}
+
+type Rule struct {
+	Project     string
+	Policy      EnforcementPolicy
+	DomainGlobs []string
+}
+
+type Decision struct {
+	Reason  string
+	Default bool
+	Result  DecisionResult
+	Project string
+}
+
+func New(logger *logrus.Logger, loader Loader, disabledActions []string) (*ACL, error) {
+	acl, err := loader.Load()
+	if err != nil {
+		return nil, err
+	}
+
+	acl.DisablePolicies(disabledActions)
+
+	err = acl.Validate()
+	if err != nil {
+		return nil, err
+	}
+
+	acl.Logger = logger
+
+	if acl.DefaultRule == nil {
+		acl.Warn("no default rule set")
+	}
+	return acl, nil
+}
+
+// Add associates a rule with the specified service after verifying the rule's
+// policy and domains are valid.
+func (acl *ACL) Add(svc string, r Rule) error {
+	err := acl.PolicyDisabled(svc, r.Policy)
+	if err != nil {
+		return err
+	}
+
+	err = acl.ValidateDomains(r.DomainGlobs)
+	if err != nil {
+		return err
+	}
+
+	acl.Rules[svc] = r
```

Comment:
Is there ever possible that we would overwrite a pre-existing rule here? If so, is that something we should warn or error on?

Reply:
Nice catch. Will fix this.

Reply:
Nice fix, thanks for the additional test coverage!